Networking Introduction Book Chapter (#917)
quinn-rs/quinn: docs/book/src/networking-introduction.md

# Networking Introduction

In this chapter, you will find a very short introduction to various networking concepts. 
Those concepts are key for understanding the motivations for QUIC. 

## 1. TCP/IP and UDP Comparison

> **TCP:** stands for 'transmission control protocol' and adds certain guarantees ontop of [IP][3]. 

Author: Good one. I doubted it as well. My editor did not like it either. I'll change it

> **TCP:** stands for 'transmission control protocol' and adds certain guarantees ontop of [IP][3]. 

Author: Ah: on top of/atop of/atop is correct english

> **TCP:** stands for 'transmission control protocol' and adds certain guarantees ontop of [IP][3]. 

Reviewer: I did some further research and "stands for" does appear to be correct English, so feel free to leave this if you prefer.

> It forms the backbone for almost everything you do online, from web browsing to IRC to email to file transfer.

Author: I removed it

> The protocol is [reliable ordered][guarantees] in nature.

Author: I will merge TCP/UDP and guarantees section together.

**UDP** stands for 'user datagram protocol' and adds certain guarantees ontop of [IP][3], but unlike TCP, 
instead of adding lots of features and complexity, UDP is a very thin layer over IP and is also [unreliable][guarantees] in nature.

| Feature |  TCP  | UDP | QUIC
| :-------------: | :-------------: | :-------------: | :-------------: |
|  [Connection-Oriented][6]           |       Yes                      | No                       | Yes
|  [Transport Guarantees][guarantees] | [Reliable Ordered][guarantees] | [Unreliable][guarantees] | [Reliable Ordered][guarantees] and [Unreliable][guarantees] 
|  Packet Transfer                    | [Stream-based][4]              | Message based            | Message based and Stream based
|  Automatic [fragmentation][8]       | Yes                            | Yes, but better to stay below datagram size limit | Yes
|  Header Size                        |  20 bytes                      | 8 bytes                  |  16 bytes  

> |  [Control Flow, Congestion Avoidance/Control][5] | Yes               | No                       |  Yes, and user controlled                                          

Author: I merged the two chapters into one

## 2. The 5 Transport Guarantees

> There are 5 main different ways you can transfer data:

Author: I removed 3 of the 2 guarantees keeping only unreliable and reliable and defining the properties of those. I merged them into the TCP/UDP section,.

> | Transport Guarantees         | Packet Drop [(1)][1]  | Packet Duplication [(2)][2] | Packet Order | Packet Delivery |

Author: Sounds good to me

| :-------------:              | :-------------: | :-------------: | :-------------: | :-------------:
|   **Unreliable Unordered**   |       Any       |      Yes        |     No          |   No
|   **Unreliable Sequenced**   |    Any + old    |      No         |     Sequenced   |   No
|   **Reliable Unordered**     |       No        |      No         |     No          |   Yes
|   **Reliable Ordered**       |       No        |      No         |     Ordered     |   Yes
|   **Reliable Sequenced**     |    Only old     |      No         |     Sequenced   |   Only newest

Unreliability gives great uncertainty with a lot of freedom, while reliability gives great certainty with costs in speed and freedom.

> That is why protocols such as QUIC, RUDP, SCTP, QUIC, netcode, laminar are built on UDP instead of TCP. 

Author: Good. Tho, one remark, is it good to go into this in a later section? Like questions, #916 are common, in my opinion, we should spend some time comparing them. Also in amethyst, benvy, game networking channels I repeatedly saw this issue.

UDP gives the end-user more control over the transmission than TCP can do. 

## 3. Issues with TCP 

> Now the golden question: "Why choose so much uncertainty with UDP when TCP is so reliable and safe?". 

Author: Agreed. It depends on a personal writing style, however, lets go for the technical formal one 👍

To answer that question we will have to delve a little deeper into some issues with TCP. 

### Head-of-line Blocking

One of the biggest issues with TCP is that of Head-of-line blocking. 
It is a convenient feature because it ensures that all packages are sent and arrive in [order][guarantees]. 

> However, in cases of high throughput (multiplayer game networking) and big load in a short time (web page load), this can be catastrophic to your application performance.

Reviewer: This doesn't mention multiplexing yet. I think multiplexing is pretty essential to the concept of HoL blocking.

Author: See the web section

The issue is demonstrated in the following animation:

![Head of line blocking][animation] 

This animation shows that if a certain packet drops in transmission, all packets have to wait at the transport layer until it is resent by the other end.
If the dropped packet is resent and arrived then all packets are freed from the transport layer. 

Let's look at two areas where this head-of-line blocking issue is a huge deal.

**Multiplayer Game Networking**

Multiplayer action games are based on a constant stream of packets sent at a speed ranging from 10 to 30 packets per second.
For the most part, the data in these packages are so time-sensitive that only the most recent data is useful.
You can think of the input and position of the player, the orientation and speed, and the state of the physical objects in the world.
If a single packet drops out we can not afford to queue up 10-30 packets a second until the lost packet is retransmitted. 

> This could cause annoying lag behavior and a bad user experience. 

Reviewer: While QUIC might work well for game networking and this is the reason you are interested in it, I think we should mention web networking first, as this was the reason it was designed in the first place. Because it is well designed, it also benefits the types of things for game networking, but this seems almost accidental.

Author: Good idea, I had not a specific reason for it

We will take a deeper dive into this subject when looking at QUIC multiplexing.
    
### Connection Setup Duration

In the standard HTTP+TLS+TCP stack, TCP needs a handshake to establish a session between server and client, 
and TLS needs its handshake to ensure that the session is secured.

![TCP-handshake](./images/tcp-handshake.svg.png)

First, the source sends an 'SYN initial request' packet to the target server to start the dialogue. 
Then the target server sends an 'SYN-ACK packet' to agree to the process.
Lastly, the source sends an 'ACK packet' to the target to confirm the process, after which the message exchange can start. 
 
Now if we want to secure the TCP connection, we have to use a protocol like TLS on top of it. 
In the case of TLS versions older than 1.3, an additional three more handshake messages are required.

You can see how expensive it is to create a secure TCP connection. 
In a scenario of TCP and TLS 1.2 with a 100ms latency, we need to wait for 6 x 100ms = 600ms to set up a connection. 
If the website is big, an additional load time can make the website load over a second. 

> This, of course, is disturbing for our short attention spans. 

Reviewer: I'm skeptical this is a big win compared to TCP + TLS 1.3 and all the same features.

Author: I did state older than 1.3, but still with TCP + TLS 1.3 we need 4 round trips in total if I'm correct? The point I want to get at with this section is to be able to refer it later if we compare the QUIC connection setup with the TCP connection setup.

### Requests in Segment

A TCP segment can only carry a single HTTP/1.1 Request/Response. 
Consequently, a large number of small segments may be sent within

> an HTTP/1.1 session which can lead to overhead.

Reviewer: It's unclear to me what you mean here. I don't think there is a material benefit for QUIC over TCP here.

Author: Agree, after reviewing I dont see much value in it as well

### Client Connection Initiation

HTTP/1.1 transfers are always initiated by the client. 
This decreases the performance of HTTP/1.1 significantly when loading embedded files because a server has to
wait for a request from the client, even if the server knows

> that the client needs a specific resource.

Reviewer: This seems just wrong. The text seems to talk about what's known as server push, which (a) HTTP 2 also has and (b) was recently disabled in Chrome because it doesn't appear to work well in practice.

Author: Agree, after reviewing I dont see much value in it as well

# Networking Introduction

In this chapter, you will find a very short introduction to various networking concepts. 
These concepts are important to understanding when to use QUIC.

## 1. TCP/IP and UDP Comparison

Let's compare TCP, UDP, and QUIC.

- **unreliable**: Transport packeten zijn er niet van verzekerd dat ze aan komen bij hun eindbestemming en op volgorde zijn. 
- **reliable**: Transport packeten zijn er van verzekerd dat ze op volgorde op hun eindbestemming komen.

Unreliability gives great uncertainty with a lot of freedom, while reliability gives great certainty with costs in speed and freedom.
That is why some protocols such as QUIC, SCTP are built on UDP instead of TCP. 

| Feature |  TCP  | UDP | QUIC
| :-------------: | :-------------: | :-------------: | :-------------: |
|  [Connection-Oriented][6]           |       Yes         | No                       | Yes
|  Transport Guarantees               | Reliable          | Unreliable               | Reliable and/or unreliable with extension 

> |  Packet Transfer                    | [Stream-based][4] | Message based            | Message based and/or Stream based

Author: What about unreliable datagram transport?

> |  Packet Transfer                    | [Stream-based][4] | Message based            | Message based and/or Stream based

Reviewer: That seems like an implementation detail. At some level all of these are implemented on top of IP, which is unreliable as well, but that has little bearing on how they get used.

> |  Packet Transfer                    | [Stream-based][4] | Message based            | Message based and/or Stream based

Author: 👍

|  Header Size                        |  20 bytes         | 8 bytes                  |  ~16 bytes(depending on connection id)  

> |  [Control Flow, Congestion Avoidance/Control][5] | Yes  | No                       |  ** Yes, and possible controlled by userspace                                          

Author: I removed it and kept the footnote we had written for it last time.

It has several difficulties that QUIC tries to resolve. 

### Head-of-line Blocking

One of the biggest issues with TCP is that of Head-of-line blocking. 
It is a convenient feature because it ensures that all packages are sent and arrive in order. 
However, in cases of high throughput (multiplayer game networking) and big load in a short time (web page load), this can severely impact latency.

The issue is demonstrated in the following animation:

![Head of line blocking][animation] 

This animation shows that if a certain packet drops in transmission, all packets have to wait at the transport layer until it is resent by the other end.
If the dropped packet is resent and arrived then all packets are freed from the transport layer. 

Let's look at two areas where this head-of-line blocking issue creates issues. 

**Web Networking**

The last years websites have been growing in size which causes. 
This increases the loading time and makes head-of-line blocking a more concerning topic. 
To tackle this issue HTTP-2 has introduced a technique called multiplexing. 
In short, this means that multiple TCP streams are initialized to communicate with the server. 

> It allows a server to transfer multiple sources in parallel over a single stream.

Reviewer: As websites increasingly need a larger number of HTTP requests (HTML, CSS, JavaScript, images) to display all content, the impact of head-of-line blocking has also increased. To improve on this, HTTP 2 introduced request multiplexing within a TCP data stream, which allows servers to stream multiple responses at the same time. However, data loss of a single packet will still block all response streams because they exist within the context of a single TCP stream.

**Multiplayer Game Networking**

> The web space is not the only area where this head-of-line blocking is a major concern.

Reviewer: I still don't think it makes sense to talk about this here, since head-of-line blocking isn't really what causes the problems with datagrams these kinds of games need. Additionally "reliability" here seems to have an entirely different meaning than what has been discussed before.

Author: Hrm. What entirely different meaning do you see? Reliability and head of line blocking are two concepts interwoven concepts don't you think? Simply said, a drop of a single packet causes buffering which is not plausible in games.

Reviewer: Reliability conceptually is about "am I sure the other side is going to see this data". Head-of-line blocking has more to do with ordering. In the world of TCP vs UDP, reliability and ordering are codependent, but conceptually I don't think that makes sense. If you needed reliability for a game transport, you could build a UDP-based transport protocol that does away with ordering, and it would have reliability without head-of-line blocking concerns.

Author: Okay, so we can remove it then?

Multiplayer action games work with a constant flow of packets sent at an interval ranging between 10 to 30 packets per second.
For the most part, the data in these packets is so time sensitive that only the most recent data can be used. 
Therefore, it cannot be afforded to queue 10-30 packets per second until the lost packet is resent.
Most multiplayer network solutions build a custom protocol on top of UDP to address head-of-line blocking issues while maintaining reliability.
   
### Connection Setup Duration

> In the usual HTTP+TLS+TCP stack, TCP needs 6 handshake messages to set up a session between server and client, 
> and TLS needs its own handshake to make sure the session is secure.  
> This handshake consists of 6 messages for TLS 1.2 or lower, and 4 messages for setting up the 'initial' connection over TLS 1.3.
> Despite the '0-RTT' function of TLS 1.3, which allows you to resume a previous connection in 0-RTT, the 6 TCP handshake messages are still required.
> Nevertheless, QUIC has the 0-RTT feature as well and is not restrained by the 6 TCP handshake messages.
> This implies that QUIC is able to provide encrypted true 0-RTT connections. 

Reviewer: ```suggestion
In the usual TCP + TLS + HTTP stack, TCP needs 6 handshake messages to set up a session between server and client. TLS performs its own, sending 4 messages for setting up an initial connection over TLS 1.3. By integrating the transport protocol and TLS handshakes, QUIC can make connection setup more efficient.
```

# Networking Introduction

In this chapter, you will find a very short introduction to various networking concepts. 
These concepts are important to understanding when to use QUIC.

## 1. TCP/IP and UDP Comparison

> Lets define some transport guarantees and their properties:

Reviewer: ```suggestion
Let's define some transport guarantees and their properties:
```

| Transport Guarantees | Packet Loss [(1)][1]  | Packet Duplication [(2)][2] | Packet Order | Packet Delivery |
| :-------------:   | :-------------: | :-------------: | :-------------: | :-------------:
|   **Unreliable**  |       Any       |      Yes        |     No          |   No

> |   **Reliable**    |       No        |      No         |     Ordered     |   Yes

Reviewer: Still not a fan of this table -- I feel it conflates too many concepts. I'd prefer a concise bullet list explanation of unreliable, unordered (UDP) and reliable, ordered (TCP).

Author: In this bullet point list, do you still want to mention those points?

Reviewer: I think only loss and order matter -- these are related; the other things fall out of these.

Now, lets compare TCP, UDP, and QUIC.

| Feature |  TCP  | UDP | QUIC
| :-------------: | :-------------: | :-------------: | :-------------: |
|  [Connection-Oriented][6]           |       Yes         | No                       | Yes
|  Transport Guarantees              | Reliable Ordered  | Unreliable               | Reliable Ordered and Unreliable 
|  Packet Transfer                    | [Stream-based][4] | Message based            | Message based and/or Stream based
|  Automatic [fragmentation][8]       | Yes               | Yes, ip-fragmentation    | Yes
|  Header Size                        |  20 bytes         | 8 bytes                  |  16 bytes  
|  [Control Flow, Congestion Avoidance/Control][5] | Yes  | No                       |  Yes, and user controlled                                          

> |  Based On | [IP][3]                | [IP][3]            |  UDP

Reviewer: Let's merge the transport guarantees/packet transfer rows here. I think we should probably just skip the fragmentation row, it doesn't seem very important. The header size for QUIC is not 16 bytes, it really depends on the length of the chosen connection IDs. In the control flow/congestion avoidance row, I believe the Yes/No distinction between TCP and UDP is pretty unhelpful, and "user controlled" seems inaccurate. I think you mean here that most QUIC implementations will run in userspace compared to TCP running control flow/congestion stuff in kernelspace, but there might be kernel implementations of QUIC in the future, and "user" is not a very clear representation of "userspace".

Author: Let's merge the transport guarantees/packet transfer rows here.  I am not sure how this should be done. Do we want to keep:  Packet Loss/Packet Duplication/Packet Order/Packet Delivery? Or reduce it unreliable/reliable? and if so do we need to we still want to mention the first mentioned points?

Reviewer: Reducing it to unreliable and reliable is probably the right way forward.

Unreliability gives great uncertainty with a lot of freedom, while reliability gives great certainty with costs in speed and freedom.
That is why some protocols such as QUIC, SCTP are built on UDP instead of TCP. 
UDP gives the end-user more control over the transmission than TCP can do. 
While QUIC is build on top of UDP it does provides the same and even more features than TCP.

## 2. Issues with TCP 

While TCP has been around for long it does have some issues that QUIC tries to solve.

### Head-of-line Blocking

One of the biggest issues with TCP is that of Head-of-line blocking. 
It is a convenient feature because it ensures that all packages are sent and arrive in order. 

> However, in cases of high throughput (multiplayer game networking) and big load in a short time (web page load), this can be catastrophic to your application performance.

Reviewer: ```suggestion
However, in cases of high throughput (multiplayer game networking) and big load in a short time (web page load), this can severely impact latency.
```

As websites get bigger and attention decreases, we need faster loading times for websites.

To tackle this issue, HTTP-2 introduced a technique called multiplexing. 
In short, this means that multiple TCP streams are initialized to communicate with the server. 
Then If one of them blocks the whole website can continue to load seemingly while that single stream is retransmitting.

We will take a deeper dive into this subject when looking at QUIC multiplexing.

**Multiplayer Game Networking**

The web space is not the only area where this head-of-line blocking is a big issue.
Multiplayer action games are based on a constant stream of packets sent at a speed ranging from 10 to 30 packets per second.
For the most part, the data in these packages are so time-sensitive that only the most recent data is useful.
You can think of the input and position of the player, the orientation and speed, and the state of the physical objects in the world.
If a single packet drops out we can not afford to queue up 10-30 packets a second until the lost packet is retransmitted. 
This could cause annoying lag behavior and a bad user experience. 
   
### Connection Setup Duration

> In the standard HTTP+TLS+TCP stack, TCP needs a handshake to establish a session between server and client, 

Reviewer: I don't think it makes sense to compare to TCP + TLS 1.2 here, so I only think this section is useful if there are improvements relative to a TCP + TLS 1.3 scenario -- otherwise, there's no strong reason to use QUIC.

Author: I rewrote it and included 1.3 in the reasoning.
